File: src/main/julia/AoC2015_06.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2015_06

include("aoc.jl")

const REGEXP = r"([ a-z]+) ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)"

function solve(input, lights, f)
    for line ∈ input
        m = match(REGEXP, line)
        start_x, end_x, start_y, end_y = [parse(Int, x) + 1
                                          for x ∈ m.captures[2:5]]
        command = m.captures[1]
    f(lights, command, start_x:start_y, end_x:end_y)
    end
    return sum(lights)
end

function part1(input)
    function process1(lights, command, start, _end)
        if command == "turn on"
            lights[start, _end] .= true
        elseif command == "turn off"
            lights[start, _end] .= false
        else
            lights[start, _end] .= .!lights[start, _end]
        end
    end
    return solve(input, falses(1000, 1000), process1)
end

function part2(input)
    function process2(lights, command, start, _end)
        if command == "turn on"
            lights[start, _end] .+= 1
        elseif command == "turn off"
            lights[start, _end] .-= 1
            lights[lights .< 0] .= 0
        else
            lights[start, _end] .+= 2
        end
    end
    return solve(input, zeros(Int, 1000, 1000), process2)
end

TEST1 = @aoc_splitlines("turn on 0,0 through 999,999")
TEST2 = @aoc_splitlines("toggle 0,0 through 999,0")
TEST3 = @aoc_splitlines("turn off 499,499 through 500,500")
TEST4 = @aoc_splitlines("turn on 0,0 through 0,0")
TEST5 = @aoc_splitlines("toggle 0,0 through 999,999")

function samples()
    @assert part1(TEST1) == 1_000_000
    @assert part1(TEST2) == 1000
    @assert part1(TEST3) == 0
    @assert part2(TEST4) == 1
    @assert part2(TEST5) == 2_000_000
end

end # module AoC2015_06

aoc_main(@__FILE__, ARGS, 2015, 6)

File: src/main/cpp/2015/06/AoC2015_06.cpp
#include <assert.h>

#include <numeric>
#include <regex>
#include <string>
#include <vector>

#include "../../aoc/aoc.hpp"
#include "../../aoc/grid/grid.hpp"

using namespace std;

const string REGEX = "([ a-z]+) ([0-9]+),([0-9]+) through ([0-9]+),([0-9]+)";

void forEachCell(int (&grid)[1'000'000], const Cell& start, const Cell& end,
                 const function<int(int)>& f) {
    for (int r = start.row() * 1000; r <= end.row() * 1000; r += 1000) {
        for (int i = r + start.col(); i <= r + end.col(); i++) {
            grid[i] = f(grid[i]);
        }
    }
}

int solve(
        const vector<string>& input,
        const function<int(int)>& turn_on,
        const function<int(int)>& turn_off,
        const function<int(int)>& toggle
) {
    int grid[1'000'000] = {};
    const regex re(REGEX);
    for (const string& line : input) {
        smatch sm;
        regex_search(line, sm, re);
        const string command = sm.str(1);
        const Cell start = Cell::at(stoi(sm.str(2)), stoi(sm.str(3)));
        const Cell end = Cell::at(stoi(sm.str(4)), stoi(sm.str(5)));
        if (command == "turn on") {
            forEachCell(grid, start, end, turn_on);
        } else if (command == "turn off") {
            forEachCell(grid, start, end, turn_off);
        } else if (command == "toggle") {
            forEachCell(grid, start, end, toggle);
        }
    }
    return accumulate(begin(grid), end(grid), 0, plus<int>());
}

int part1(const vector<string>& input) {
    return solve(
        input,
        [](int v) { return 1; },
        [](int v) { return 0; },
        [](int v) { return v == 0 ? 1 : 0; }
    );
}

int part2(const vector<string>& input) {
    return solve(
        input,
        [](int v) { return v + 1; },
        [](int v) { return v == 0 ? 0 : v - 1; },
        [](int v) { return v + 2; }
    );
}

const vector<string> TEST1 = {"turn on 0,0 through 999,999"};
const vector<string> TEST2 = {"toggle 0,0 through 999,0"};
const vector<string> TEST3 = {"turn off 499,499 through 500,500"};
const vector<string> TEST4 = {"turn on 0,0 through 0,0"};
const vector<string> TEST5 = {"toggle 0,0 through 999,999"};

void samples() {
    assert(part1(TEST1) == 1000000);
    assert(part1(TEST2) == 1000);
    assert(part1(TEST3) == 0);
    assert(part2(TEST4) == 1);
    assert(part2(TEST5) == 2000000);
}

MAIN(2015, 6)
